Feat/fortgot password+create new password flow

OmarQussas opened the pull request.

Commit:
(feat): create the auth layout

OmarQussas committed:
add/ setup for the forgot-password

- fixed a  few bugs generated after merging with dev
- created setup for the forgot-password

OmarQussas committed:
add/ forgot-password-page-implementations

- implemented the forgot password ui
- created the server action
- created a reusable where the logic lies

OmarQussas committed:
Merge remote-tracking branch 'origin/feat/otp-verification' into feat/Fortgot-Password+Create-New-Password-Flow

OmarQussas committed:
Merge remote-tracking branch 'origin/feat/register' into feat/Fortgot-Password+Create-New-Password-Flow
Integraring the Layout with the rest of the app

OmarQussas committed:
merge/ otp and layout code

- merged my current code with the layout and the otp code to complete the flow

OmarQussas committed:
add/reset-password and forgot-password

added reset password and forgot password functions flow
- server action for each one
- reusable hook to save the logic
- rendered them as components and had them ready to be conditionally rendered

OmarQussas committed:
add/ reusable component

- added reusable component for link

OmarQussas committed:
fix/minimal fixes

OmarQussas committed:
restrucutre/ created steps folder

- created the steps folder to be easily integrated with the layout

MohamedAbdelhakem17 commented on src/app/[locale]/(auth)/forgot-password/_components/steps/forgot-password-step.tsx:
It’s better to use a **unified Form component** from  [shadcn](https://ui.shadcn.com/docs/components/form) UI library instead of setting up `react-hook-form` separately in every form. That way, all forms have a consistent look and error handling, and it’s much easier to maintain or update later — any design or validation change happens in one place, keeping everything cleaner and aligned with the project’s Design System.

MohamedAbdelhakem17 commented on src/app/[locale]/(auth)/forgot-password/_components/steps/reset-password-step.tsx:
It’s recommended to use the **Form** component from the shadcn library instead of building forms manually with `useForm`, `Label`, and `Input` each time, as mentioned before.
This ensures **consistency in design**, simplifies error handling, and keeps the form logic cleaner.

Also, it’s better to **define the component’s types** (like props) in the same file when the component is small and not reused elsewhere.
This makes the codebase easier to read and maintain.

MohamedAbdelhakem17 commented on src/app/[locale]/(auth)/forgot-password/_components/steps/forgot-password-step.tsx:
It’s better to use a **unified Form component** from  [shadcn](https://ui.shadcn.com/docs/components/form) UI library instead of setting up `react-hook-form` separately in every form. That way, all forms have a consistent look and error handling, and it’s much easier to maintain or update later — any design or validation change happens in one place, keeping everything cleaner and aligned with the project’s Design System.

MohamedAbdelhakem17 commented on src/lib/schemes/auth.schemes.ts:
the same comment in  forgot password schema

Shamsmedhat replied:
The correct way to create a schema to translate the content.

Example:

```
export const useLoginSchema = () => {
  // Translation
  const t = useTranslations();

  return z.object({
    email: z.string({ required_error: t("email-required") }).email({
      message: t("email-invalid"),
    }),
    password: z
      .string({ required_error: t("password-required") })
      .min(1, t("password-required")),
  });
};

export type LoginFields = z.infer<ReturnType<typeof useLoginSchema>>;
```

MohamedAbdelhakem17 commented on src/lib/schemes/forgot-password.schema.ts:
This part can be **refactored** later to reuse the validation logic from the **register schema** once it’s implemented.
Instead of redefining the same password rules here, you could use `z.object().pick({ password: true })`  to ensure consistency and reduce code duplication.

MohamedAbdelhakem17 reviewed (commented):
Good and solid work — really well done and great effort 👏 Keep it up!

As a general improvement, it would be better to **organize all auth-related types** inside a single file, e.g. `src/lib/types/auth.d.ts`, instead of having a separate file for each small type.
The same applies to your schemas — it’s cleaner to **group all authentication-related schemas** in one place, such as `src/lib/schemas/auth.schema.ts`.

This makes the structure more maintainable, easier to navigate, and keeps all auth logic consistent and centralized.

OmarQussas committed:
fix/ made the code cleaner

OmarQussas committed:
fix/ used the form from shadcn

- used shadCN forms
- restructured the folder structures

MoaazHail commented on src/app/[locale]/(auth)/_components/_layout/authentication-heading.tsx:
```diff
@@ -0,0 +1,96 @@
+  return (
+    <h2
+      className={cn(
+        'text-zinc-800 dark:text-zinc-50 font-semibold text-2xl   ',
```
the AuthenticationHeading.title style is not like the design. you must make a font family ' Imperial Script ' and import it from ' font/google '.

MoaazHail commented on src/app/[locale]/(auth)/_components/_layout/authentication-heading.tsx:
This is awesome, but it is very complicated 👀💔.
the comments is very useful, but still the component complex 😂

Shamsmedhat replied:
This is a compound components pattern, so there is no problem with all team members learning it.

MoaazHail commented on src/app/[locale]/(auth)/_components/_layout/authentication-link.tsx:
Great one ✨💪
The component is simple and very readable.

MoaazHail commented on src/app/[locale]/(auth)/forgot-password/_hooks/use-forgot-password.ts:
Nice work! 💪✨

MoaazHail commented on src/app/[locale]/(auth)/layout.tsx:
```diff
@@ -1,7 +1,59 @@
+        <div className='w-[406px] mx-auto'>
+          <div className='flex justify-end'>
+            {/* language switcher button */}
+            <Button
+              variant='ghost'
+              size='sm'
+              className='text-zinc-700 hover:text-maroon-600 p-0 h-auto'
+            >
+              العربية
+            </Button>
```
In my opinion, you can create a new component for these lines of code because it is not a button ( navigate to route or create an action ) it is a language toggle.

MoaazHail commented on src/components/shared/password-input.tsx:
Great one ✨💪 Really clean and easy to understand.

OmarQussas replied:
Thanks ✨
Really appreciate your review 🙏

MoaazHail reviewed (commented):
Overall it is very good ❤️✨
keep going 💪

Shamsmedhat commented on src/app/[locale]/(auth)/forgot-password/_components/_shared/auth-redirect.tsx:
```diff
@@ -0,0 +1,28 @@
+  className = '',
+}) => {
+  return (
+    <div className={`flex justify-center items-center mt-4 ${className}`}>
```
We use the cn utility whenever we need to write JavaScript logic inside a className.

Shamsmedhat commented on src/app/[locale]/(auth)/forgot-password/_components/_shared/auth-redirect.tsx:
```diff
@@ -0,0 +1,28 @@
+'use client';
```
Always make sure to leave a blank line after each section in the file, such as after the "use client" directive, import statements, constants, or hooks. 

This improves readability, helps with code scanning, and makes it easier for others to maintain and understand your code structure.

Shamsmedhat commented on src/app/[locale]/(auth)/forgot-password/_components/_shared/auth-redirect.tsx:
```diff
@@ -0,0 +1,28 @@
+};
+
+export const AuthRedirect: FC<AuthRedirectProps> = ({
+  message = "Don't have an account yet?",
```
Where is the translation of this content?

Shamsmedhat replied:
Implement a default message with translation.

Shamsmedhat commented on src/app/[locale]/(auth)/forgot-password/_components/steps/reset-password-step.tsx:
```diff
@@ -0,0 +1,93 @@
+import { ResetPasswordInputs, ResetPasswordFormProps } from '@/lib/types/auth';
+import { resetPasswordSchema } from '@/lib/schemes/auth.schemes';
+
+export function ResetPasswordForm({
```
You need to translate the content.

Shamsmedhat commented on src/app/[locale]/(auth)/forgot-password/_hooks/use-forgot-password.ts:
```diff
@@ -0,0 +1,39 @@
+
+export function useForgotPassword() {
+  // initialize cookie hook
+  const { setCookieValue } = useCookie({
```
No need to store the email in cookie, you can just store it in state (memory/app) instead using cookies.

OmarQussas replied:
I would love to discuss this point further with you in a future stand-up

Shamsmedhat commented on src/app/[locale]/(auth)/forgot-password/_hooks/use-forgot-password.ts:
```diff
@@ -0,0 +1,39 @@
+
+  // main mutation
+  const mutationFn = async (email: string): Promise<ForgotPasswordResponse> => {
+    const formData = new FormData();
```
No need for this.

Shamsmedhat commented on src/app/[locale]/(auth)/forgot-password/_hooks/use-forgot-password.ts:
This is an example of the flow.

**Form component body**

```
export default function RegisterForm() {
  // Translation
  const t = useTranslations();

  // Navigation
  const router = useRouter();

  // Hooks
  const registerSchema = useRegisterSchema();
  const { isPending, error, register } = useRegister();

  // Form
  const form = useForm<RegistrationFields>({
    resolver: zodResolver(registerSchema),
    defaultValues: {
      firstName: "",
      lastName: "",
      username: "",
      email: "",
      phone: "",
      password: "",
      rePassword: "",
    },
  });

  // Functions
  function onSubmit(values: RegistrationFields) {
    register(values);
  }
```

**Field**

```
   {/* First name */}
            <FormField
              control={form.control}
              name="firstName"
              render={({ field }) => (
                <FormItem>
                  {/* Label */}
                  <FormLabel>{t("firstname-label")}</FormLabel>

                  {/* Field */}
                  <FormControl>
                    <Input
                      placeholder={t("firstname-placeholder")}
                      {...field}
                      autoComplete="given-name"
                    />
                  </FormControl>

                  {/* Feedback */}
                  <FormMessage />
                </FormItem>
              )}
            />
```

**Hook**

```
export default function useRegister() {
  // Navigation
  const router = useRouter();
  const searchParams = useSearchParams();

  // Mutation
  const { isPending, error, mutate } = useMutation({
    mutationFn: async (fields: RegistrationFields) =>
      await registerAction(fields),
     // Don't forget to handle error here
    onSuccess: () => {
      // Redirect to the the login page upon successful registration
      router.push(`/auth/login?${searchParams.toString()}`);
    },
  });

  return { isPending, error, register: mutate };
}
```

**Action**

```
"use server";

import { JSON_HEADER } from "@/lib/constants/api.constant";
import { RegistrationFields } from "@/lib/schemes/auth.schema";
import { RegisterResponse } from "@/lib/types/auth";

export const registerAction = async (
  registrationFields: RegistrationFields
) => {
  const respones = await fetch(`${process.env.API}/auth/signup`, {
    method: "POST",
    body: JSON.stringify(registrationFields),
    headers: {
      ...JSON_HEADER,
    },
  });

  const payload: APIResponse<RegisterResponse> = await respones.json();

  return payload;
};
```

Shamsmedhat reviewed (changes requested):
Nice effort and good job overall 👏,

OmarQussas committed:
feat/ translations

- added translations for forgot-password and reset-password components
- added translations for the zod schema

OmarQussas committed:
Merge branch 'dev' into feat/Fortgot-Password+Create-New-Password-Flow

OmarQussas committed:
Fix/ duplicated success message

OmarQussas committed:
feat/ redirect after new password creation
